DCSProcessor

chiarazampolli opened the pull request:
This is an example of how to process the DCS data coming from DCS. For now, one should simulate the Data Points with a macro, and then process them from there. 
More discussions will be needed to understand how to interface the DCS FLP to the standard processing workflow (with DPL).

--> Workflow to simulate and process DCS data added.

--> update of CCDB added: o2-dcs-dcs-data-workflow --max-timeframes 100 -b | o2-calibration-ccdb-populator-workflow --ccdb-path localhost:8080 -b

--> default behaviour: first TF will process the full map with all DPs; from the 2nd TF on, only a delta map is processed, DP by DP, looping on detectors that "subscribed" to that DP. 
In order to process the full delta map, option "--process-full-delta-map" should be used

-------------------

DCS processor example

example macro

Update.

Fixes + check on entries in map

Using unordered_map find, adding other types.

Not yet complete.

Fix for string type (thanks to Ruben!) and adding binary type.

Implementing reading of flags plus other changes.

* new Init taking full map;
* function to check presence of alias.

Removing commented out line

Fix indentation

For the sake of clang happiness.

chiarazampolli committed:
DCSProcessor

DCS processor example

example macro

Update.

Fixes + check on entries in map

Using unordered_map find, adding other types.

Not yet complete.

Fix for string type (thanks to Ruben!) and adding binary type.

Implementing reading of flags plus other changes.

* new Init taking full map;
* function to check presence of alias.

Removing commented out line

Fix indentation

For the sake of clang happiness.

Following comments by Ruben.

Some fixes for comments - but does not compile

Fixing specialized templates.

Clang-format.

Changing layout of templated functions + making some functions virtual

clang-format

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,122 @@
+  void processBinaries();
+  uint64_t processFlag(uint64_t flag, const char* alias);
+
+  void doSimpleMovingAverage(int nelements, std::deque<int>& vect, float& avg, bool& isSMA);
```
should not this be templated wrt ``deque``?

chiarazampolli replied:
Yes, this I can indeed template. It started more like an example.

chiarazampolli replied:
Done

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+
+//ClassImp(o2::dcs::DCSProcessor);
+
+void DCSProcessor::init(std::vector<DPID> aliaseschars, std::vector<DPID> aliasesints, std::vector<DPID> aliasesdoubles,
```
the alias arguments should be ``const std::vector<DPID>& aliaseschars`` etc.

chiarazampolli replied:
True, done

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+  for (auto i = 0; i < aliasesbinaries.size(); i++) {
+    mLatestTimestampbinaries[i] = 0;
+  }
+}
```
you would not need these loops if you would resize as e.g. ``mLatestTimestampbools.resize(aliasesbools.size(),0);``

chiarazampolli replied:
Done, thanks.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+
+//______________________________________________________________________
+
+void DCSProcessor::init(std::vector<DPID> aliases)
```
idem, ``const std::vector<DPID>& aliases``. And the same comment for initializations in the loop. But why do we need 2 alternative ``init()``?

chiarazampolli replied:
Done. 
I think that the one with only 1 argument should be enough, but for testing, maybe one wants to use the second. If you prefer, I can remove the one with multiple vectors.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+
+  // first, we need to check if there are the Data Points that we need
+
+  auto foundChars = 0, foundInts = 0, foundDoubles = 0, foundUInts = 0,
```
in such cases better use explicit ``int``

chiarazampolli replied:
Done

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+  // char type
+  auto s = mAliaseschars.size();
+  if (s > 0) {
+    for (size_t i = 0; i != mAliaseschars.size(); ++i) {
```
``for (size_t i = 0; i != s; ++i) {``

chiarazampolli replied:
Yes of course.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+    for (size_t i = 0; i != mAliaseschars.size(); ++i) {
+      int count = 0;
+      std::unordered_map<DPID, DPVAL>::const_iterator it;
+      processAlias(mAliaseschars[i], DeliveryType::RAW_CHAR, map, count, it);
```
instead of providing the iterator as an argument and additionally count to check its validity, better to do
````
auto processAlias(DPID& alias, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map)
{
  // processing basic checks for map: all needed aliases must be present
  auto it = map.find(alias);
  if (it == map.end()) {
    LOG(ERROR) << "Element " << alias << " not found " << std::endl;
    count = 0;
  }
  DeliveryType tt = alias.get_type();
  if (tt != type) {
    LOG(FATAL) << "Delivery Type of alias " << alias.get_alias() << " does not match definition in DCSProcessor (" << type << ")! Please fix";
  }
  return it;
}
````
and check its validity as
````
auto it = processAlias(mAliaseschars[i], DeliveryType::RAW_CHAR, map);
if (it==map.end()) {
  continue;
}
````

chiarazampolli replied:
Indeed I can check the validity where I call "processAlias", I did not see it since I was taking a piece of code and making it a separate function.
As for return object vs argument, what is the difference? I mean, they do the same job. Is one better?
As for using auto, I would explicitly mention the type for clarity.

chiarazampolli replied:
BTW, I cannot return it because it is a local variable. This is why I did it in the other way. Am I wrong?

shahor02 replied:
The difference is just shorter and more clear code, the outcome is the same. You can return a local variable (it will be moved), you cannot do this with pointers or refs on local vars.

chiarazampolli replied:
Ok, i thought it would have been copied, which is why I would have used a reference, but one cannot use it to a local variable. Since it is moved, it is of course fine.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+          mLatestTimestampints[i] = etime;
+        }
+      }
+    }
```
should not we put this to separate function, instead of repeating the same code many times?

chiarazampolli replied:
I thought about it, but the map that I fill is a different one every time (mDpsintsmap, mDpscharsmap...). Maybe I can template it.

chiarazampolli replied:
Hello @shahor02 , 
I tried to template this part, which works, but needs an explicit specialization for string and binary types. But when compiling, I get errors, and I don't really understand how to fix them. First, I got the error: 
```
error: explicit specialization of 'processArrayType' in class scope
```
So I moved the specialization outside the class, but still it does not work. I never used specializations of templates, do you know what I am doing wrong?
Thanks!
Chiara

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+
+//______________________________________________________________________
+
+void DCSProcessor::processAlias(DPID& alias, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map, int& count, std::unordered_map<DPID, DPVAL>::const_iterator& it)
```
see above

chiarazampolli replied:
See above. I removed the "count" because this check is now where I call the function, but the argument "it" is still there.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,634 @@
+
+//______________________________________________________________________
+
+void DCSProcessor::processChars()
```
these methods also could be templated, e.g.

````
template<typename T>
void DCSProcessor::processType(const std::vector<DPID>& aliases, std::unordered_map<DPID, T> &data)
{
  for (size_t i = 0; i != aliases.size(); ++i) {
    auto& id = aliases[i];
    auto& valvec = data[id];
    LOG(INFO) << "size = " << valvec.size();
    for (size_t j = 0; j < valvec.size(); j++) {
      LOG(INFO) << "DP = " << aliases[i] << " , value[" << j << "] = " << valvec[j];
    }
  }
}
````

chiarazampolli replied:
The reason for not making it templated is that the detectors might have a different way to process ints, chars, bools... My implementation does not do anything and is always the same, but it is just an example.

shahor02 reviewed (commented):
Hi @chiarazampolli 
please see a few comments below.

shahor02 commented:
If custom behaviour for some type is needed, the template can be specialized.

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,143 @@
+class DCSProcessor
+{
+
+  using Ints = std::vector<int>;
+  using Chars = std::vector<char>;
+  using Doubles = std::vector<double>;
+  using Binaries = std::array<uint64_t, 7>;
+  using Strings = std::array<char, 56>;
+
+  using DQChars = std::deque<char>;
+  using DQInts = std::deque<int>;
+  using DQDoubles = std::deque<double>;
+  using DQUInts = std::deque<uint32_t>;
+  using DQBools = std::deque<bool>;
+  using DQStrings = std::deque<Strings>;
+  using DQTimes = std::deque<uint32_t>;
+  using DQBinaries = std::deque<Binaries>;
+
+  using DPID = o2::dcs::DataPointIdentifier;
+  using DPVAL = o2::dcs::DataPointValue;
+  using DPCOM = o2::dcs::DataPointCompositeObject;
```
please move this to ``public`` part

chiarazampolli replied:
Done, good point.

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,143 @@
+
+  int process(const std::unordered_map<DPID, DPVAL>& map);
+
+  template<typename T>
+  int processArrayType(const std::vector<DPID>& array, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map, std::vector<uint64_t>& latestTimeStamp, std::unordered_map<DPID, T>& destmap);
```
The template definition should be in the header file, either in the class declaration body or after it, as inline methods are usually defined, i.e.
````
template<typename T>
int DCSProcessor::processArrayType(....) {
   // processing the array of type T
   int found = 0;
   // implementation you have currently in cxx file 
}
````

chiarazampolli replied:
Not sure I understand the comment: I declared the specialized templated function in the header, but the body is in the cxx.

shahor02 replied:
no, you have the non-specialized template implementation in the cxx, while it should be ``.h``

chiarazampolli replied:
Ah, the non-specialized!
But are you sure it is necessary? Like it is now, it works (compiles + macro works).

shahor02 replied:
In general, it is necessary, in your case it works because at the moment these methods are used only in the same compilation module where they are defined. But you will not be able to call them as class methods from some another module.

chiarazampolli replied:
Hello Ruben,

Changed according to our discussion.

Chiara

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,143 @@
+  using DPVAL = o2::dcs::DataPointValue;
+
+  //template<>
+  int processArrayType(const std::vector<DPID>& array, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map, std::vector<uint64_t>& latestTimeStamp, std::unordered_map<DPID, DQStrings>& destmap);
```
You did not indicate the class name, should be 
````
template<>
int DataProcessor::processArrayType(const std::vector<DPID>& array, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map, std::vector<uint64_t>& latestTimeStamp, std::unordered_map<DPID, DQStrings>& destmap);
````
Then, in the cxx you can keep the specializations you have now.

chiarazampolli replied:
TRUE!!!!

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,143 @@
+  ClassDefNV(DCSProcessor, 0);
+};
+
+  using Ints = std::vector<int>;
+  using Chars = std::vector<char>;
+  using Doubles = std::vector<double>;
+  using Binaries = std::array<uint64_t, 7>;
+  using Strings = std::array<char, 56>;
+
+  using DQStrings = std::deque<Strings>;
+  using DQBinaries = std::deque<Binaries>;
+
+  using DPID = o2::dcs::DataPointIdentifier;
+  using DPVAL = o2::dcs::DataPointValue;
```
you don't need to redefine these, just use below e.g. ``DCSProcessor::DQStrings`` instead of ``DQStrings``

chiarazampolli replied:
Yes, after making the public, it works of course.

shahor02 reviewed (commented):
Hi @chiarazampolli 
I put some comments in the code.

chiarazampolli commented:
Thanks @shahor02 , now the specialized templated function work :-)

chiarazampolli commented:
Hello @shahor02 , 

Concerning this:

> If custom behaviour for some type is needed, the template can be specialized.

about the "processChars/Ints/etc": I am not convinced that it should be templated, even with specialized templates. This is the part that will be detector-specific and data point specific. 

BTW, I would make part of the functions virtual, what do you think? In case detector-specific implementations want to derive for this class. 

Chiara

chiarazampolli commented:
Hello @ktf , @ihrivnac ,

It is a bit complicated to see what fails in clang-format. But I think it is not related to my code (I did run clang-format, and see: https://github.com/chiarazampolli/AliceO2/pull/21/files#diff-8f6abb8e8f8cc2b3b328e85a810057d2). Also, the o2 build fails form something that I do not understand. And gpu failure is also unrelated to my PR.

Right?

This PR is not yet ready to be merged, but I would like to understand the outcome of the checks.

Chiara

shahor02 commented:
> about the "processChars/Ints/etc": I am not convinced that it should be templated, even with specialized templates. This is the part that will be detector-specific and data point specific.
> BTW, I would make part of the functions virtual, what do you think? In case detector-specific implementations want to derive for this class.

OK, it was just a suggestion, I don't think all detectors will implement all flavours of processArrayType, and also the virtual method (which is fine here) can be templated. So, up to you.

chiarazampolli commented:
> > about the "processChars/Ints/etc": I am not convinced that it should be templated, even with specialized templates. This is the part that will be detector-specific and data point specific.
> > BTW, I would make part of the functions virtual, what do you think? In case detector-specific implementations want to derive for this class.
> 
> OK, it was just a suggestion, I don't think all detectors will implement all flavours of processArrayType, and also the virtual method (which is fine here) can be templated. So, up to you.

This I kept as they were. I made virtual other functions, that I think are likely to be overwritten by the detectors.

chiarazampolli commented:
Hello,

Can someone (@ihrivnac ? @MichaelLettrich ? @ktf ?) confirm if the issue with clang-format is related to me or not? I ran it over my PR already, and in the log (very difficult to find now, it was different before), I see:
```
patching file DataFormats/simulation/include/SimulationDataFormat/Stack.h
```
not touched by me.
Chiara

chiarazampolli commented:
...also the compilation error is not related to my PR:

```
Scanning dependencies of target O2test-rans-FrequencyTable
Scanning dependencies of target O2test-rans-EncodeDecode
Scanning dependencies of target O2exe-rans-encode-decode-8
Scanning dependencies of target O2exe-rans-encode-decode-32
Scanning dependencies of target O2exe-rans-encode-decode-16
[  4%] Building CXX object Utilities/rANS/CMakeFiles/O2test-rans-EncodeDecode.dir/test/test_ransEncodeDecode.cxx.o
[  4%] Building CXX object Utilities/rANS/CMakeFiles/O2test-rans-FrequencyTable.dir/test/test_ransFrequencyTable.cxx.o
[  4%] Building CXX object Utilities/rANS/CMakeFiles/O2exe-rans-encode-decode-8.dir/run/bin-encode-decode.cxx.o
[  4%] Building CXX object Utilities/rANS/CMakeFiles/O2exe-rans-encode-decode-32.dir/run/bin-encode-decode.cxx.o
[  4%] Building CXX object Utilities/rANS/CMakeFiles/O2exe-rans-encode-decode-16.dir/run/bin-encode-decode.cxx.o
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
gmake[2]: *** [Utilities/DataCompression/CMakeFiles/O2test-datacompression-dc_primitives.dir/test/test_dc_primitives.cxx.o] Error 4
gmake[1]: *** [Utilities/DataCompression/CMakeFiles/O2test-datacompression-dc_primitives.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
gmake[2]: *** [DataFormats/MemoryResources/CMakeFiles/O2test-memoryresources-MemoryResources.dir/test/testMemoryResources.cxx.o] Error 4
gmake[1]: *** [DataFormats/MemoryResources/CMakeFiles/O2test-memoryresources-MemoryResources.dir/all] Error 2
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
gmake[2]: *** [Utilities/rANS/CMakeFiles/O2test-rans-EncodeDecode.dir/test/test_ransEncodeDecode.cxx.o] Error 4
gmake[1]: *** [Utilities/rANS/CMakeFiles/O2test-rans-EncodeDecode.dir/all] Error 2
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
gmake[2]: *** [Utilities/rANS/CMakeFiles/O2exe-rans-encode-decode-8.dir/run/bin-encode-decode.cxx.o] Error 4
gmake[1]: *** [Utilities/rANS/CMakeFiles/O2exe-rans-encode-decode-8.dir/all] Error 2
[  4%] Linking CXX executable ../../stage/tests/o2-test-rans-FrequencyTable
[  4%] Built target O2test-rans-FrequencyTable
[  4%] Linking CXX executable ../../stage/bin/o2-rans-encode-decode-32
[  4%] Linking CXX executable ../../stage/bin/o2-rans-encode-decode-16
[  4%] Built target O2exe-rans-encode-decode-32
[  4%] Built target O2exe-rans-encode-decode-16
gmake: *** [all] Error 2
## End sw/BUILD/O2-latest/log
```

chiarazampolli committed:
Adding workflow for DCS processing

Fixing workflow.

INFO --> DEBUG

clang...

beautification

Forgot to delete pointer

Fixing compilation

shahor02 commented:
Hi @chiarazampolli 

You should have added ``#pragma link C++ class std::unordered_map<o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue> + ;`` to ``DetectorsDCSLinkDef.h``, but I see this is also not sufficient: the DPL does not know how the transmit the unordered_map.

@ktf could you comment how ``unordered_map`` should be messaged? 
I've tried also to wrap it to a class with dictionary:
````
  class DCSMap : public std::unordered_map<o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue>
  {
    public:
    using std::unordered_map<o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue>::unordered_map;
    
    ClassDefNV(DCSMap,1);
  };
````
in my private branch https://github.com/shahor02/AliceO2/tree/dcs_tmp, commit https://github.com/shahor02/AliceO2/commit/9b05e962694740655dff0da4e35d4bb7b12eea7c
but then I get:
````
[100%] Building CXX object Detectors/DCS/CMakeFiles/O2exe-dcs-dcs-data-workflow.dir/testWorkflow/dcs-data-workflow.cxx.o
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h: In instantiation of ‘decltype(auto) o2::framework::InputRecord::get(R, int) const [with T = o2::dcs::DCSMap*; R = const char*]’:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h:90:46:   required from here
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:367:88: error: ambiguous template instantiation for ‘struct o2::framework::has_root_dictionary<o2::dcs::DCSMap, void>’
                          (is_messageable<typename std::remove_pointer<T>::type>::value || has_root_dictionary<typename std::remove_pointer<T>::type>::value)) {
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/SerializationMethods.h:16:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataRefUtils.h:15,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TypeTraits.h:120:8: note: candidates are: template<class T> struct o2::framework::has_root_dictionary<T, typename std::conditional<false, o2::framework::class_member_checker<decltype (declval<T>().Class())>, void>::type> [with T = o2::dcs::DCSMap]
 struct has_root_dictionary<
        ^~~~~~~~~~~~~~~~~~~~
   T,
   ~~    
   std::conditional_t<
   ~~~~~~~~~~~~~~~~~~~
     false,
     ~~~~~~
     class_member_checker<
     ~~~~~~~~~~~~~~~~~~~~~
       decltype(std::declval<T>().Class())>,
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     void>> : public std::true_type {
     ~~~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TypeTraits.h:137:7: note:                 template<class T> class o2::framework::has_root_dictionary<T, typename std::enable_if<o2::framework::is_container<T>::value>::type> [with T = o2::dcs::DCSMap]
 class has_root_dictionary<T, typename std::enable_if<is_container<T>::value>::type>
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:367:88: error: incomplete type ‘o2::framework::has_root_dictionary<o2::dcs::DCSMap, void>’ used in nested name specifier
                          (is_messageable<typename std::remove_pointer<T>::type>::value || has_root_dictionary<typename std::remove_pointer<T>::type>::value)) {
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:400:109: error: use of ‘static auto o2::framework::DataRefUtils::as(const o2::framework::DataRef&) [with T = o2::framework::ROOTSerialized<o2::dcs::DCSMap, void>]’ before deduction of ‘auto’
         std::unique_ptr<ValueT const, Deleter<ValueT const>> result(DataRefUtils::as<ROOTSerialized<ValueT>>(ref).release());
                                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:426:94: error: use of ‘static auto o2::framework::DataRefUtils::as(const o2::framework::DataRef&) [with T = o2::framework::ROOTSerialized<o2::dcs::DCSMap*, void>]’ before deduction of ‘auto’
         std::unique_ptr<T const, Deleter<T const>> result(DataRefUtils::as<ROOTSerialized<T>>(ref).release());
                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: error: inconsistent deduction for auto return type: ‘std::unique_ptr<const o2::dcs::DCSMap, o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap> >’ and then ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >’
         return result;
                ^~~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: error: no matching function for call to ‘std::unique_ptr<const o2::dcs::DCSMap, o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap> >::unique_ptr(std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >)’
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/shahoian/alice/sw/ubuntu1804_x86-64/FairMQ/v1.4.23-1/include/fairmq/FairMQMessage.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TMessageSerializer.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/RootSerializationSupport.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataRefUtils.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/usr/include/c++/7/bits/unique_ptr.h:255:2: note: candidate: template<class _Up, class> std::unique_ptr<_Tp, _Dp>::unique_ptr(std::auto_ptr<_Up>&&)
  unique_ptr(auto_ptr<_Up>&& __u) noexcept;
  ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:255:2: note:   template argument deduction/substitution failed:
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: note:   ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >’ is not derived from ‘std::auto_ptr<_Up>’
         return result;
                ^~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/shahoian/alice/sw/ubuntu1804_x86-64/FairMQ/v1.4.23-1/include/fairmq/FairMQMessage.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TMessageSerializer.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/RootSerializationSupport.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataRefUtils.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/usr/include/c++/7/bits/unique_ptr.h:247:2: note: candidate: template<class _Up, class _Ep, class> std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Up, _Ep>&&)
  unique_ptr(unique_ptr<_Up, _Ep>&& __u) noexcept
  ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:247:2: note:   template argument deduction/substitution failed:
/usr/include/c++/7/bits/unique_ptr.h:233:7: note: candidate: std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Tp, _Dp>&&) [with _Tp = const o2::dcs::DCSMap; _Dp = o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap>]
       unique_ptr(unique_ptr&& __u) noexcept
       ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:233:7: note:   no known conversion for argument 1 from ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >’ to ‘std::unique_ptr<const o2::dcs::DCSMap, o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap> >&&’
/usr/include/c++/7/bits/unique_ptr.h:228:12: note: candidate: template<class _Up, class> constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr(std::nullptr_t)
  constexpr unique_ptr(nullptr_t) noexcept : _M_t() { }
            ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:228:12: note:   template argument deduction/substitution failed:
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: note:   cannot convert ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >(result)’ (type ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >’) to type ‘std::nullptr_t’
         return result;
                ^~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/shahoian/alice/sw/ubuntu1804_x86-64/FairMQ/v1.4.23-1/include/fairmq/FairMQMessage.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TMessageSerializer.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/RootSerializationSupport.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataRefUtils.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/usr/include/c++/7/bits/unique_ptr.h:219:7: note: candidate: std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Tp, _Dp>::pointer, typename std::remove_reference<_To>::type&&) [with _Tp = const o2::dcs::DCSMap; _Dp = o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap>; std::unique_ptr<_Tp, _Dp>::pointer = const o2::dcs::DCSMap*; typename std::remove_reference<_To>::type = o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap>]
       unique_ptr(pointer __p,
       ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:219:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/7/bits/unique_ptr.h:207:7: note: candidate: std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Tp, _Dp>::pointer, typename std::conditional<std::is_reference<_Dp>::value, _Dp, const _Dp&>::type) [with _Tp = const o2::dcs::DCSMap; _Dp = o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap>; std::unique_ptr<_Tp, _Dp>::pointer = const o2::dcs::DCSMap*; typename std::conditional<std::is_reference<_Dp>::value, _Dp, const _Dp&>::type = const o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap>&]
       unique_ptr(pointer __p,
       ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:207:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/7/bits/unique_ptr.h:196:2: note: candidate: template<class _Up, class> std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Tp, _Dp>::pointer)
  unique_ptr(pointer __p) noexcept
  ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:196:2: note:   template argument deduction/substitution failed:
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: note:   cannot convert ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >(result)’ (type ‘std::unique_ptr<o2::dcs::DCSMap* const, o2::framework::InputRecord::Deleter<o2::dcs::DCSMap* const> >’) to type ‘std::unique_ptr<const o2::dcs::DCSMap, o2::framework::InputRecord::Deleter<const o2::dcs::DCSMap> >::pointer {aka const o2::dcs::DCSMap*}’
         return result;
                ^~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/shahoian/alice/sw/ubuntu1804_x86-64/FairMQ/v1.4.23-1/include/fairmq/FairMQMessage.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/TMessageSerializer.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/RootSerializationSupport.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataRefUtils.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:14,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/usr/include/c++/7/bits/unique_ptr.h:183:12: note: candidate: template<class _Up, class> constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr()
  constexpr unique_ptr() noexcept
            ^~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:183:12: note:   template argument deduction/substitution failed:
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:13:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/InputRecord.h:427:16: note:   candidate expects 0 arguments, 1 provided
         return result;
                ^~~~~~
In file included from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/ProcessingContext.h:14:0,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/AlgorithmSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataProcessorSpec.h:13,
                 from /home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:11:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataAllocator.h: In instantiation of ‘void o2::framework::DataAllocator::snapshot(const o2::framework::Output&, const T&) [with T = o2::dcs::DCSMap]’:
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataAllocator.h:425:81:   required from ‘auto o2::framework::DataAllocator::snapshot(o2::framework::OutputRef&&, Args&& ...) [with Args = {o2::dcs::DCSMap&}]’
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h:108:72:   required from here
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataAllocator.h:303:56: error: incomplete type ‘o2::framework::has_root_dictionary<o2::dcs::DCSMap, void>’ used in nested name specifier
     } else if constexpr (has_root_dictionary<T>::value == true || is_specialization<T, ROOTSerialized>::value == true) {
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/shahoian/alice/sw/SOURCES/O2/v1.2.0/0/Framework/Core/include/Framework/DataAllocator.h:343:7: error: static assertion failed: data type T not supported by API, 
 specializations available for
 - trivially copyable, non-polymorphic structures
 - std::vector of messageable structures or pointers to those
 - types with ROOT dictionary and implementing ROOT ClassDef interface
       static_assert(always_static_assert_v<T>,
       ^~~~~~~~~~~~~
````

chiarazampolli commented:
> Hi @chiarazampolli
> 
> You should have added `#pragma link C++ class std::unordered_map<o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue> + ;` to `DetectorsDCSLinkDef.h`, but I see this is also not sufficient: the DPL does not know how the transmit the unordered_map.
> 

> 
Hello @shahor02 ,

Of course... Sorry for overlooking it, when I saw the error, I just thought unorderd_map were not passable, but indeed, one needs to specify them in the LinkDef. Thanks for taking a look and going even further. Let's see what @ktf suggests.

Cheers,

Chiara

ihrivnac commented:
> Hello @ktf , @ihrivnac ,
> 
> It is a bit complicated to see what fails in clang-format. But I think it is not related to my code (I did run clang-format, and see: https://github.com/chiarazampolli/AliceO2/pull/21/files#diff-8f6abb8e8f8cc2b3b328e85a810057d2). Also, the o2 build fails form something that I do not understand. And gpu failure is also unrelated to my PR.
> 
> Right?
> 
> This PR is not yet ready to be merged, but I would like to understand the outcome of the checks.
> 
> Chiara

Hi @chiarazampolli ,

the clang-format check was added recently by @ktf  to help developers with agreed code formatting. It should check only the files modified in the PR. Please, ignore this check as experimental for the time being, until this is fixed.

chiarazampolli commented:
Hello @shahor02 , @ktf , 

With the workaround by Giulio to make the unordered_map messageable (without the problem of the root dictionary), now everything seems to work. 
For me, unless you have other additional comments, the PR can be merged.  

Thanks for the help.

Chiara

shahor02 reviewed (dismissed):
OK, will merge once CIs passed.

chiarazampolli commented:
Sorry @shahor02 , there was a delete missing... Added now.

shahor02 commented:
Hi @chiarazampolli 
You need to add ``#include <unordered_map>`` to ``dcs-data-workflow.cxx``. Also, please change ``<string.h>`` to ``<string>`` in 
``DCSProcessor.cxx``

chiarazampolli committed:
Sending binary buffer.

The unorderd_map is created in the processor part.

bug fix

clang...

chiarazampolli commented:
> you should copy from rawchar, not dh:
> 
> ```
> memcpy(&dptmp, rawchar + i *sizeof(DPCOM), sizeof(DPCOM));
> ```

No comment... :-)

chiarazampolli commented:
Error from clang-format seems unrelated:
```
++ git-clang-format-8 --commit b899de32c27bbced540aff635f1dd17c8020d5a9 --diff --binary /usr/bin/clang-format-8 Analysis/Core/include/Analysis/PairCuts.h Analysis/DataModel/include/Analysis/CFDerived.h Analysis/DataModel/include/Analysis/SecondaryVertexHF.h Analysis/Tasks/PWGCF/CMakeLists.txt Analysis/Tasks/PWGCF/correlations.cxx Analysis/Tasks/PWGCF/correlationsFiltered.cxx Analysis/Tasks/PWGCF/filterCF.cxx Analysis/Tasks/PWGHF/CMakeLists.txt Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx Analysis/Tasks/PWGHF/taskDPlus.cxx Analysis/Tasks/centralityTable.cxx Analysis/Tasks/eventSelection.cxx DataFormats/Detectors/ITSMFT/ITS/include/DataFormatsITS/TrackITS.h DataFormats/Detectors/TPC/include/DataFormatsTPC/LaserTrack.h DataFormats/Reconstruction/CMakeLists.txt DataFormats/Reconstruction/include/ReconstructionDataFormats/Track.h DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackLTIntegral.h DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackParametrization.h DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackParametrizationWithError.h DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackUtils.h DataFormats/Reconstruction/src/Track.cxx DataFormats/Reconstruction/src/TrackParametrization.cxx DataFormats/Reconstruction/src/TrackParametrizationWithError.cxx Detectors/DCS/CMakeLists.txt Detectors/DCS/include/DetectorsDCS/DCSProcessor.h Detectors/DCS/include/DetectorsDCS/DataPointValue.h Detectors/DCS/src/DCSProcessor.cxx Detectors/DCS/test/processor_dpcom_o2.C Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h Detectors/DCS/testWorkflow/dcs-data-workflow.cxx Detectors/FIT/FV0/simulation/include/FV0Simulation/Digitizer.h Detectors/FIT/FV0/simulation/include/FV0Simulation/FV0DigParam.h Detectors/FIT/FV0/simulation/src/Digitizer.cxx Detectors/ITSMFT/ITS/tracking/src/Tracker.cxx Detectors/TOF/base/include/TOFBase/WindowFiller.h Detectors/TOF/calibration/include/TOFCalibration/CalibTOFapi.h Detectors/TOF/calibration/src/CalibTOF.cxx Detectors/TOF/calibration/src/CalibTOFapi.cxx Detectors/TOF/simulation/include/TOFSimulation/Digitizer.h Detectors/TOF/simulation/src/Digitizer.cxx Framework/Core/include/Framework/ASoA.h Framework/Core/include/Framework/AnalysisDataModel.h Framework/Core/src/ASoA.cxx Framework/Core/src/DataAllocator.cxx Framework/Core/test/test_ASoA.cxx Generators/share/external/GenCosmics.C Generators/share/external/GenCosmicsLoader.C Generators/share/external/README.md Steer/DigitizerWorkflow/src/FV0DigitizerSpec.cxx Utilities/DataSampling/CMakeLists.txt Utilities/DataSampling/src/DataSampling.cxx Utilities/DataSampling/test/test_DataSampling.cxx Utilities/DataSampling/test/test_DataSamplingEmpty.json cmake/O2RootMacroExclusionList.cmake run/CMakeLists.txt run/o2sim_parallel.cxx
fatal: ambiguous argument 'Analysis/Core/include/Analysis/PairCuts.h': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
+ RESULT_OUTPUT=
Error: Process completed with exit code 2.
```

chiarazampolli committed:
Stressing the simulation using 50K points

chiarazampolli committed:
Template of simple moving average.

chiarazampolli committed:
Adding publishing in CCDB part

e.g.:
o2-dcs-dcs-data-workflow --max-timeframes 100 -b | o2-calibration-ccdb-populator-workflow --ccdb-path localhost:8080 -b

clang-format

Setting TF

Clang-format

Generating and receiving also a delta map

(Not yet processed)

Clang...

Fix to stupid mistake

shahor02 commented:
Hi @chiarazampolli 

CI reports a real error, better use ``uint64_t payload[7];`` here, no need to delete.
````
Error while processing /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx.
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h:229:5: error: 'delete' applied to a pointer that was allocated with 'new[]'; did you mean 'delete[]'? [clang-diagnostic-mismatched-new-delete]
    delete payload;
    ^
          []
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h:115:25: note: allocated with 'new[]' here
    uint64_t* payload = new uint64_t[7];
````

chiarazampolli commented:
> Hi @chiarazampolli
> 
> CI reports a real error, better use `uint64_t payload[7];` here, no need to delete.
> 
> ```
> Error while processing /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/dcs-data-workflow.cxx.
> /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h:229:5: error: 'delete' applied to a pointer that was allocated with 'new[]'; did you mean 'delete[]'? [clang-diagnostic-mismatched-new-delete]
>     delete payload;
>     ^
>           []
> /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/DCS/testWorkflow/DCSDataGeneratorSpec.h:115:25: note: allocated with 'new[]' here
>     uint64_t* payload = new uint64_t[7];
> ```

Yes, it was a stupid mistake, sorry, and I did not check the CI in detail, since it was taking long and in principle I want to still add something (processing of delta map). 

Chiara

shahor02 commented on Detectors/DCS/testWorkflow/dcs-data-workflow.cxx:
```diff
@@ -0,0 +1,43 @@
+
+#include "DetectorsDCS/DataPointIdentifier.h"
+#include "DetectorsDCS/DataPointValue.h"
+#include "Framework/TypeTraits.h"
```
The CI wants you to ``#include <unordered_map>`` here

chiarazampolli replied:
Hello @shahor02 , It is included. I can move it above, let's see if it makes it happy. I will do it when ready with few minor changes.

chiarazampolli replied:
Let's see further tests...

chiarazampolli committed:
Including processing of delta map.

Clang

chiarazampolli committed:
Implementing ProcessDP.

chiarazampolli committed:
Possibility to process per Data Point only needed detectors.

And other changes...

Clang

chiarazampolli committed:
De-templating some functions

so that detectors can override them if deriving from DCSProcessor.

clang

shahor02 commented on Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h:
```diff
@@ -0,0 +1,244 @@
+
+    TStopwatch s;
+    LOG(INFO) << "TF: " << tfid << " -->  receiving binary data...";
+    s.Start();
+    auto rawchar = pc.inputs().get<const char*>("input");
+    s.Stop();
+    LOG(INFO) << "TF: " << tfid << " -->  ...binary data received: realTime = " << s.RealTime() << ", cpuTime = " << s.CpuTime();
+    LOG(INFO) << "TF: " << tfid << " -->  receiving (delta) binary data...";
+    s.Start();
+    auto rawcharDelta = pc.inputs().get<const char*>("inputDelta");
+    s.Stop();
```
timing of such a small intervals is impossible to measure (you are resetting the timer at every start). Usually I define data member timers which are triggered by ``Start(true);``  / ``Stop()`` around relatively time consuming pieces of code 
 and accumulated measurement in many TFs.

chiarazampolli replied:
Yes, the resetting is intentional. Indeed the blocks that I check are very fast. I can implement the summing over more TFs, but the most important timing (for the processing) is ok also at TF level (at least from the numbers I see, it can be measured).

chiarazampolli replied:
Improvements on timing implemented.

shahor02 commented on Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h:
```diff
@@ -0,0 +1,244 @@
+        bool resetStopwatch = true;
+        LOG(INFO) << "TF: " << tfid << " -->  starting (delta) processing for detector...";
+        for (const auto& dpcom : dcsmapDelta) {
+          std::vector<int> detVect = mDetectorAlias[dpcom.first];
+          //for (int idet = 0; idet < detVect.size(); idet++) {
```
should not you indeed loop over entries of ``detVect[i]`` ?

chiarazampolli replied:
Yes, it was just for test, as I had defined 2 detectors, but I have the processor for one only. Now I deinfe only 1, so I can use the correct loop.

shahor02 commented on Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h:
```diff
@@ -0,0 +1,244 @@
+        LOG(INFO) << "TF: " << tfid << " -->  ...processing (delta) in detector loop done: realTime = " << s.RealTime() << ", cpuTime = " << s.CpuTime();
+        // now preparing CCDB object
+        //for (int idet = 0; idet < detVect.size(); idet++) {
+        for (int idet = 0; idet < 1; idet++) { // for now I test only 1 DCS processor
```
why the ccdb object preparation is called only in per-DP loop?

chiarazampolli replied:
In case we pass the full map, the CCDB object is created in the processMap method directly. I cannot do like that if I process 1 DP at a time. Do you prefer that I create it in the Data Processor also when I process the full map, for consistency with the case of processing per DP?

shahor02 replied:
I see, ok. I hope in real life we will not call ``prepareCCDBobject`` for every TF / map update. But as a demo it is fine.

chiarazampolli replied:
I hope so too :-) As you said, it is a demo.

shahor02 commented on Detectors/DCS/src/DCSProcessor.cxx:
```diff
@@ -0,0 +1,553 @@
+
+  // first we check if the DP is in the list for the detector
+  if (type == DeliveryType::RAW_CHAR) {
+    auto it = std::find(mAliaseschars.begin(), mAliaseschars.end(), dpid);
```
this search is actually redundant (as we discussed), it will be much faster to use ``unordered_map<alias, index>`` to check if the processed needs this alias.

chiarazampolli replied:
The idea here is to see if this alias was actually really defined for the DCSProcessor. If I create the unordered_map that you propose, it is true that I can check if an alias should be processed or skipped with the "[]", but if I don't use find but "[]", then in case the alias is not there (--> there is a misconfiguration), it will be created. This is not what I want. I want to check that the configuration of the DCSProcessor matches with the list of aliases that in the DataProcessor it is expected to process. 

I can add the unordered_map that you propose, but with a different meaning --> checking if the DP (which somewhere else was confirmed to be in the detector configuration) needs to be indeed processed, or should be skipped (even though I would not know why, to decide to skip it, one would need to read it, so process it). 

Do yo agree?

shahor02 replied:
I mean the unordered_map exactly for that purpose: to check (using find, which is mush faster for the map than for the vector) if the DP is requested by this processor.
Additionally, the value associated with the key can be used as a slot to fill (up to detector)

shahor02 commented on Detectors/DCS/include/DetectorsDCS/DCSProcessor.h:
```diff
@@ -0,0 +1,299 @@
+
+  if (vect.size() <= nelements) {
+    //avg = std::accumulate(vect.begin(), vect.end(), 0.0) / vect.size();
+    avg = (avg * (vect.size() - 1) + vect[vect.size() - 1]) / vect.size();
```
use ``vect.back()`` instead of ``vect[vect.size() - 1]``.

chiarazampolli replied:
Done, thanks

shahor02 commented on Detectors/DCS/testWorkflow/DCSDataProcessorSpec.h:
```diff
@@ -0,0 +1,244 @@
+  void run(o2::framework::ProcessingContext& pc) final
+  {
+    auto tfid = o2::header::get<o2::framework::DataProcessingHeader*>(pc.inputs().get("input").header)->startTime;
+    mDCSproc.setTF(tfid);
```
here you use ``mDCSproc``, but then  actual data filling and ccdb preparation is done on mDCSprocVect

chiarazampolli replied:
I forgot to set it for mDCSprocVect. Fixed. 
For clarity: mDCSproc is there to benchmark in case we do not process per timestamp.

shahor02 reviewed (commented):
Hi Chiara,

Please see some comments below.

chiarazampolli committed:
Improving timing evaluation.

clang.

clang 2

clang 3

Again clang

Clang!!!!

Clang!!!

Maybe like this clang will be happy

shahor02 reviewed (commented):
Hi @chiarazampolli 

see below, let me know if you want me to merge this PR as is, speeding up per-DP processing using ``unordered_map::find`` can be done later.

chiarazampolli committed:
Using unordered_map instead of vectors.

clang

chiarazampolli committed:
Revert "Using unordered_map instead of vectors."

This reverts commit 40ce3e718f268e50d5574e8287c6694b6c26355a.

chiarazampolli committed:
Adding map to keep track of needed aliases, to be used in find.

Stopping TStopwatches before starting measuring

clang

chiarazampolli commented:
Hello @shahor02 ,
Yesterday all tests were green, I don't know why GPU is not ok now, but it seems unrelated.
Apart from that, I think we can merge (at your convenience).

Thanks!

Chiara

shahor02 reviewed (approved):
GPU CI fails due to the usual memory overrun in the compiler, merging.

ktf commented:
For the record, the GPU builder should now be ok for new PRs. Please nag again if not the case.
